### ISSUE TITLE
Using KLDiv with CMNRL

### ISSUE THREAD

**Issue by yjoonjang** [UKPLab/sentence-transformers#3484]
Hi @tomaarsen,
We can see that in the [E5 paper](https://arxiv.org/abs/2212.03533), the KLDivLoss is used with InfoNCE Loss.

$\min D_{KL}(p_{ce}, p_{stu}) + \alpha L_{cont}$

Is combining two losses available in sentence-transformers? 

Thank you.

- Youngjoon Jang

**Comment by tomaarsen** [comment]
Hello!

Yes, although not fully directly. In short, in v5.0 we added `compute_loss_from_embeddings` to a bunch of the loss functions. You can create a new loss function that initializes multiple ones, and calls this `compute_loss_from_embeddings` on each. For example:

```python
from typing import Iterable
import torch
from sentence_transformers.losses import MultipleNegativesRankingLoss, DistillKLDivLoss

class InfoNCEKLDivLoss(torch.nn.Module):
    def __init__(self, model, info_nce_scale=20.0, kl_div_temperature=1.0, gather_across_devices=False, info_nce_weight=1.0, kl_div_weight=1.0):
        super().__init__()
        self.model = model
        self.info_nce = MultipleNegativesRankingLoss(model, scale=info_nce_scale, gather_across_devices=gather_across_devices)
        self.kl_div = DistillKLDivLoss(model, temperature=kl_div_temperature, gather_across_devices=gather_across_devices)
        self.info_nce_weight = info_nce_weight
        self.kl_div_weight = kl_div_weight

    def forward(self, sentence_features: Iterable[dict[str, torch.Tensor]], labels: torch.Tensor) -> torch.Tensor:
        embeddings = [self.model(sentence_feature)["sentence_embedding"] for sentence_feature in sentence_features]

        # Feel free to return a dictionary of losses if you want to compute multiple losses,
        # they will be logged separately as well as summed up, and the sum will be used for backpropagation.
        loss = {
            "info_nce": self.info_nce.compute_loss_from_embeddings(embeddings, labels) * self.info_nce_weight,
            "kl_div": self.kl_div.compute_loss_from_embeddings(embeddings, labels) * self.kl_div_weight,
        }
        return loss
```

However, do recognize that different losses require different formats. See e.g. the [Loss Overview](https://sbert.net/docs/sentence_transformer/loss_overview.html).
* `DistillKLDivLoss` requires anchor, positive, 1+ negatives with a score for each positive/negative compared to the anchor.
* `MultipleNegativesRankingLoss` (InfoNCE) requires anchor, positive, 0+ negatives, and ignores any score.

So you can use them together, but you will have to use anchor, positive, 1+ negatives with a score for each positive/negative compared to the anchor. You can mine for this with `mine_hard_negatives` if you want. Notably, the `n-tuple-scores` format matches this immediately:

<img width="1019" height="302" alt="Image" src="https://github.com/user-attachments/assets/f41a3411-1edc-4b5b-bc97-be8f5369dfe8" />

Personally I would recommend using `mine_hard_negatives` with a small embedding model and `n-tuple-scores`, saving the output dataset, and then using a reranker model on the saved dataset to update the scores from embedding model scores to reranker model scores, but it's all up to you.

- Tom Aarsen

**Comment by yjoonjang** [comment]
Thank you so much for your response !!
